compile error: duration.inl - duration literals are not found

seems like the duration literals are not resolved during compilation of iceoryx_hoofs/internal/units/duration.inl in vs

e.g iceoryx_hoofs/internal/units/duration.inl(79,23): error C3861: '_ns': identifier not found

they are defined later in the header (inline namespace duration_literals)

@niclar with `/std:c++latest` I am drowning in compiler errors right from the first file:
```
  file.cpp
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,52
): error C2760: syntax error: '&' was unexpected here; expected 'expression' [C:\Users\me\Development\iceoryx\build\hoo
fs\platform\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,52
): error C2187: syntax error: '&' was unexpected here [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_pla
tform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,1)
: error C2059: syntax error: '>' [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,63
): error C2988: unrecognizable template declaration/definition [C:\Users\me\Development\iceoryx\build\hoofs\platform\ic
eoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,63
): error C2059: syntax error: '>' [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(52,1)
: error C2143: syntax error: missing ';' before '{' [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platf
orm.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(52,1)
: error C2447: '{': missing function header (old-style formal list?) [C:\Users\me\Development\iceoryx\build\hoofs\platf
orm\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\source\file.cpp(25,9): error C3861: '__Win32Call': identifie
r not found [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platform.vcxproj]
  fnctl.cpp
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,52
): error C2760: syntax error: '&' was unexpected here; expected 'expression' [C:\Users\me\Development\iceoryx\build\hoo
fs\platform\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,52
): error C2187: syntax error: '&' was unexpected here [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_pla
tform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,1)
: error C2059: syntax error: '>' [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,63
): error C2988: unrecognizable template declaration/definition [C:\Users\me\Development\iceoryx\build\hoofs\platform\ic
eoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(51,63
): error C2059: syntax error: '>' [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(52,1)
: error C2143: syntax error: missing ';' before '{' [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platf
orm.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\include\iceoryx_hoofs/platform/win32_errorHandling.hpp(52,1)
: error C2447: '{': missing function header (old-style formal list?) [C:\Users\me\Development\iceoryx\build\hoofs\platf
orm\iceoryx_platform.vcxproj]
C:\Users\me\Development\iceoryx\iceoryx_hoofs\platform\win\source\fnctl.cpp(23,1): error C3861: '__Win32Call': identifi
er not found [C:\Users\me\Development\iceoryx\build\hoofs\platform\iceoryx_platform.vcxproj]
  getopt.cpp
```

--- a/iceoryx_hoofs/include/iceoryx_hoofs/concurrent/resizeable_lockfree_queue.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/concurrent/resizeable_lockfree_queue.hpp
@@ -125,7 +125,8 @@ class ResizeableLockFreeQueue : protected LockFreeQueue<ElementType, MaxCapacity
   private:
     using BufferIndex = typename Base::BufferIndex;
     std::atomic<uint64_t> m_capacity{MaxCapacity};
-    std::atomic_flag m_resizeInProgress;
+    // must be operator= otherwise it is undefined, see https://en.cppreference.com/w/cpp/atomic/ATOMIC_FLAG_INIT
+    std::atomic_flag m_resizeInProgress = ATOMIC_FLAG_INIT;
     iox::cxx::vector<BufferIndex, MaxCapacity> m_unusedIndices;
 
     /// @brief      Increase the capacity by some value.

--- a/iceoryx_hoofs/include/iceoryx_hoofs/concurrent/resizeable_lockfree_queue.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/concurrent/resizeable_lockfree_queue.hpp
@@ -125,7 +125,7 @@ class ResizeableLockFreeQueue : protected LockFreeQueue<ElementType, MaxCapacity
   private:
     using BufferIndex = typename Base::BufferIndex;
     std::atomic<uint64_t> m_capacity{MaxCapacity};
-    std::atomic_flag m_resizeInProgress{false};
+    std::atomic_flag m_resizeInProgress;
     iox::cxx::vector<BufferIndex, MaxCapacity> m_unusedIndices;
 
     /// @brief      Increase the capacity by some value.

--- a/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
@@ -54,6 +54,20 @@ using add_const_conditionally_t = typename add_const_conditionally<T, C>::type;
 template <typename>
 constexpr bool always_false_v = false;
 
+#if __cplusplus < 201703L
+template <typename C, typename... Cargs>
+struct invoke_result
+{
+    using type = typename std::result_of<C(Cargs...)>::type;
+};
+#elif __cplusplus >= 201703L
+template <typename C, typename... Cargs>
+struct invoke_result
+{
+    using type = typename std::invoke_result<C, Cargs...>::type;
+};
+#endif
+
 ///
 /// @brief Verifies whether the passed Callable type is in fact invocable with the given arguments
 ///
@@ -63,12 +77,7 @@ struct is_invocable
     // This variant is chosen when Callable(ArgTypes) successfully resolves to a valid type, i.e. is invocable.
     /// @note result_of is deprecated, switch to invoke_result in C++17
     template <typename C, typename... As>
-#if __cplusplus < 201703L
-    static constexpr std::true_type test(typename std::result_of<C(As...)>::type*) noexcept
-    // use std::invoke_result so that iceoryx can be used in c++20 projects
-#elif __cplusplus >= 201703L
-    static constexpr std::true_type test(typename std::invoke_result<C, As...>::type*) noexcept
-#endif
+    static constexpr std::true_type test(typename cxx::invoke_result<C, As...>::type*) noexcept
     {
         return {};
     }
@@ -94,14 +103,8 @@ template <typename ReturnType, typename Callable, typename... ArgTypes>
 struct is_invocable_r
 {
     template <typename C, typename... As>
-    static constexpr std::true_type
-#if __cplusplus < 201703L
-    test(std::enable_if_t<std::is_convertible<typename std::result_of<C(As...)>::type, ReturnType>::value>*) noexcept
-    // use std::invoke_result so that iceoryx can be used in c++20 projects
-#elif __cplusplus >= 201703L
-    test(
-        std::enable_if_t<std::is_convertible<typename std::invoke_result<C, As...>::type, ReturnType>::value>*) noexcept
-#endif
+    static constexpr std::true_type test(
+        std::enable_if_t<std::is_convertible<typename cxx::invoke_result<C, As...>::type, ReturnType>::value>*) noexcept
     {
         return {};
     }

--- a/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
@@ -54,10 +54,11 @@ using add_const_conditionally_t = typename add_const_conditionally<T, C>::type;
 template <typename>
 constexpr bool always_false_v = false;
 
-#if __cplusplus < 201703L
+// windows defines __cplusplus as 199711L
+#if __cplusplus < 201703L && !defined(_WIN32)
 template <typename C, typename... Cargs>
 using invoke_result = std::result_of<C(Cargs...)>;
-#elif __cplusplus >= 201703L
+#elif __cplusplus >= 201703L || defined(_WIN32)
 template <typename C, typename... Cargs>
 using invoke_result = std::invoke_result<C, Cargs...>;
 #endif

--- a/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
@@ -56,16 +56,10 @@ constexpr bool always_false_v = false;
 
 #if __cplusplus < 201703L
 template <typename C, typename... Cargs>
-struct invoke_result
-{
-    using type = typename std::result_of<C(Cargs...)>::type;
-};
+using invoke_result = std::result_of<C(Cargs...)>;
 #elif __cplusplus >= 201703L
 template <typename C, typename... Cargs>
-struct invoke_result
-{
-    using type = typename std::invoke_result<C, Cargs...>::type;
-};
+using invoke_result = std::invoke_result<C, Cargs...>;
 #endif
 
 ///

--- a/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp
@@ -38,7 +38,7 @@ enum class TimeSpecReference
 
 class Duration;
 
-inline namespace duration_literals
+namespace duration_literals
 {
 /// @brief Constructs a new Duration object from nanoseconds
 constexpr Duration operator"" _ns(unsigned long long int) noexcept; // PRQA S 48

--- a/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl
@@ -21,46 +21,6 @@ namespace iox
 {
 namespace units
 {
-namespace duration_literals
-{
-inline constexpr Duration operator"" _ns(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromNanoseconds(value);
-}
-
-inline constexpr Duration operator"" _us(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromMicroseconds(value);
-}
-
-inline constexpr Duration operator"" _ms(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromMilliseconds(value);
-}
-
-inline constexpr Duration operator"" _s(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromSeconds(value);
-}
-
-inline constexpr Duration operator"" _m(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromMinutes(value);
-}
-
-inline constexpr Duration operator"" _h(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromHours(value);
-}
-
-inline constexpr Duration operator"" _d(unsigned long long int value) noexcept // PRQA S 48
-{
-    return Duration::fromDays(value);
-}
-
-} // namespace duration_literals
-
-
 inline constexpr Duration::Duration(const Seconds_t seconds, const Nanoseconds_t nanoseconds) noexcept
     : m_seconds(seconds)
     , m_nanoseconds(nanoseconds)
@@ -546,7 +506,43 @@ inline constexpr Duration operator*(const T& lhs, const Duration& rhs) noexcept
     return rhs * lhs;
 }
 
+namespace duration_literals
+{
+inline constexpr Duration operator"" _ns(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromNanoseconds(value);
+}
+
+inline constexpr Duration operator"" _us(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromMicroseconds(value);
+}
+
+inline constexpr Duration operator"" _ms(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromMilliseconds(value);
+}
+
+inline constexpr Duration operator"" _s(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromSeconds(value);
+}
+
+inline constexpr Duration operator"" _m(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromMinutes(value);
+}
+
+inline constexpr Duration operator"" _h(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromHours(value);
+}
+
+inline constexpr Duration operator"" _d(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromDays(value);
+}
+} // namespace duration_literals
 } // namespace units
 } // namespace iox
-
 #endif // IOX_HOOFS_UNITS_DURATION_INL

--- a/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl
@@ -1,5 +1,5 @@
 // Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
-// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl
@@ -21,6 +21,46 @@ namespace iox
 {
 namespace units
 {
+namespace duration_literals
+{
+inline constexpr Duration operator"" _ns(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromNanoseconds(value);
+}
+
+inline constexpr Duration operator"" _us(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromMicroseconds(value);
+}
+
+inline constexpr Duration operator"" _ms(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromMilliseconds(value);
+}
+
+inline constexpr Duration operator"" _s(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromSeconds(value);
+}
+
+inline constexpr Duration operator"" _m(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromMinutes(value);
+}
+
+inline constexpr Duration operator"" _h(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromHours(value);
+}
+
+inline constexpr Duration operator"" _d(unsigned long long int value) noexcept // PRQA S 48
+{
+    return Duration::fromDays(value);
+}
+
+} // namespace duration_literals
+
+
 inline constexpr Duration::Duration(const Seconds_t seconds, const Nanoseconds_t nanoseconds) noexcept
     : m_seconds(seconds)
     , m_nanoseconds(nanoseconds)
@@ -74,39 +114,83 @@ inline constexpr unsigned long long int Duration::positiveValueOrClampToZero(con
 }
 
 template <typename T>
-constexpr Duration Duration::fromNanoseconds(const T value) noexcept
+inline constexpr Duration Duration::fromNanoseconds(const T value) noexcept
 {
-    return operator"" _ns(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    auto seconds = static_cast<Duration::Seconds_t>(clampedValue / Duration::NANOSECS_PER_SEC);
+    auto nanoseconds = static_cast<Duration::Nanoseconds_t>(clampedValue % Duration::NANOSECS_PER_SEC);
+    return Duration{seconds, nanoseconds};
 }
 template <typename T>
-constexpr Duration Duration::fromMicroseconds(const T value) noexcept
+inline constexpr Duration Duration::fromMicroseconds(const T value) noexcept
 {
-    return operator"" _us(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    auto seconds = static_cast<Duration::Seconds_t>(clampedValue / Duration::MICROSECS_PER_SEC);
+    auto nanoseconds = static_cast<Duration::Nanoseconds_t>((clampedValue % Duration::MICROSECS_PER_SEC)
+                                                            * Duration::NANOSECS_PER_MICROSEC);
+    return Duration{seconds, nanoseconds};
 }
 template <typename T>
-constexpr Duration Duration::fromMilliseconds(const T value) noexcept
+inline constexpr Duration Duration::fromMilliseconds(const T value) noexcept
 {
-    return operator"" _ms(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    auto seconds = static_cast<Duration::Seconds_t>(clampedValue / Duration::MILLISECS_PER_SEC);
+    auto nanoseconds = static_cast<Duration::Nanoseconds_t>((clampedValue % Duration::MILLISECS_PER_SEC)
+                                                            * Duration::NANOSECS_PER_MILLISEC);
+    return Duration{seconds, nanoseconds};
 }
 template <typename T>
-constexpr Duration Duration::fromSeconds(const T value) noexcept
+inline constexpr Duration Duration::fromSeconds(const T value) noexcept
 {
-    return operator"" _s(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    constexpr Duration::Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Duration::Seconds_t>::max()};
+    if (clampedValue > MAX_SECONDS_BEFORE_OVERFLOW)
+    {
+        /// @todo #607 issue warning or fail
+
+        return Duration::max();
+    }
+    return Duration{static_cast<Duration::Seconds_t>(clampedValue), 0U};
 }
 template <typename T>
-constexpr Duration Duration::fromMinutes(const T value) noexcept
+inline constexpr Duration Duration::fromMinutes(const T value) noexcept
 {
-    return operator"" _m(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    constexpr uint64_t MAX_MINUTES_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / Duration::SECS_PER_MINUTE};
+    if (clampedValue > MAX_MINUTES_BEFORE_OVERFLOW)
+    {
+        /// @todo #607 issue warning or fail
+
+        return Duration::max();
+    }
+    return Duration{static_cast<Duration::Seconds_t>(clampedValue * Duration::SECS_PER_MINUTE), 0U};
 }
 template <typename T>
-constexpr Duration Duration::fromHours(const T value) noexcept
+inline constexpr Duration Duration::fromHours(const T value) noexcept
 {
-    return operator"" _h(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    constexpr uint64_t MAX_HOURS_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / Duration::SECS_PER_HOUR};
+    if (clampedValue > MAX_HOURS_BEFORE_OVERFLOW)
+    {
+        /// @todo #607 issue warning or fail
+
+        return Duration::max();
+    }
+    return Duration{static_cast<Duration::Seconds_t>(clampedValue * Duration::SECS_PER_HOUR), 0U};
 }
 template <typename T>
-constexpr Duration Duration::fromDays(const T value) noexcept
+inline constexpr Duration Duration::fromDays(const T value) noexcept
 {
-    return operator"" _d(positiveValueOrClampToZero(value, __PRETTY_FUNCTION__));
+    auto clampedValue = positiveValueOrClampToZero(value, __PRETTY_FUNCTION__);
+    constexpr uint64_t SECS_PER_DAY{Duration::HOURS_PER_DAY * Duration::SECS_PER_HOUR};
+    constexpr uint64_t MAX_DAYS_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / SECS_PER_DAY};
+    if (clampedValue > MAX_DAYS_BEFORE_OVERFLOW)
+    {
+        /// @todo #607 issue warning or fail
+
+        return Duration::max();
+    }
+    return Duration{static_cast<Duration::Seconds_t>(clampedValue * SECS_PER_DAY), 0U};
 }
 
 inline constexpr Duration::Duration(const struct timeval& value) noexcept
@@ -455,81 +539,6 @@ inline constexpr Duration Duration::operator*(const T& rhs) const noexcept
     return multiplyWith<T>(rhs);
 }
 
-inline namespace duration_literals
-{
-inline constexpr Duration operator"" _ns(unsigned long long int value) noexcept // PRQA S 48
-{
-    auto seconds = static_cast<Duration::Seconds_t>(value / Duration::NANOSECS_PER_SEC);
-    auto nanoseconds = static_cast<Duration::Nanoseconds_t>(value % Duration::NANOSECS_PER_SEC);
-    return Duration{seconds, nanoseconds};
-}
-
-inline constexpr Duration operator"" _us(unsigned long long int value) noexcept // PRQA S 48
-{
-    auto seconds = static_cast<Duration::Seconds_t>(value / Duration::MICROSECS_PER_SEC);
-    auto nanoseconds =
-        static_cast<Duration::Nanoseconds_t>((value % Duration::MICROSECS_PER_SEC) * Duration::NANOSECS_PER_MICROSEC);
-    return Duration{seconds, nanoseconds};
-}
-
-inline constexpr Duration operator"" _ms(unsigned long long int value) noexcept // PRQA S 48
-{
-    auto seconds = static_cast<Duration::Seconds_t>(value / Duration::MILLISECS_PER_SEC);
-    auto nanoseconds =
-        static_cast<Duration::Nanoseconds_t>((value % Duration::MILLISECS_PER_SEC) * Duration::NANOSECS_PER_MILLISEC);
-    return Duration{seconds, nanoseconds};
-}
-
-inline constexpr Duration operator"" _s(unsigned long long int value) noexcept // PRQA S 48
-{
-    constexpr Duration::Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Duration::Seconds_t>::max()};
-    if (value > MAX_SECONDS_BEFORE_OVERFLOW)
-    {
-        /// @todo #607 issue warning or fail
-
-        return Duration::max();
-    }
-    return Duration{static_cast<Duration::Seconds_t>(value), 0U};
-}
-
-inline constexpr Duration operator"" _m(unsigned long long int value) noexcept // PRQA S 48
-{
-    constexpr uint64_t MAX_MINUTES_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / Duration::SECS_PER_MINUTE};
-    if (value > MAX_MINUTES_BEFORE_OVERFLOW)
-    {
-        /// @todo #607 issue warning or fail
-
-        return Duration::max();
-    }
-    return Duration{static_cast<Duration::Seconds_t>(value * Duration::SECS_PER_MINUTE), 0U};
-}
-
-inline constexpr Duration operator"" _h(unsigned long long int value) noexcept // PRQA S 48
-{
-    constexpr uint64_t MAX_HOURS_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / Duration::SECS_PER_HOUR};
-    if (value > MAX_HOURS_BEFORE_OVERFLOW)
-    {
-        /// @todo #607 issue warning or fail
-
-        return Duration::max();
-    }
-    return Duration{static_cast<Duration::Seconds_t>(value * Duration::SECS_PER_HOUR), 0U};
-}
-
-inline constexpr Duration operator"" _d(unsigned long long int value) noexcept // PRQA S 48
-{
-    constexpr uint64_t SECS_PER_DAY{Duration::HOURS_PER_DAY * Duration::SECS_PER_HOUR};
-    constexpr uint64_t MAX_DAYS_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / SECS_PER_DAY};
-    if (value > MAX_DAYS_BEFORE_OVERFLOW)
-    {
-        /// @todo #607 issue warning or fail
-
-        return Duration::max();
-    }
-    return Duration{static_cast<Duration::Seconds_t>(value * SECS_PER_DAY), 0U};
-}
-
-} // namespace duration_literals
 
 template <typename T>
 inline constexpr Duration operator*(const T& lhs, const Duration& rhs) noexcept

--- a/iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/timer.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/timer.hpp
@@ -124,7 +124,8 @@ class Timer
         /// @brief the descriptor is unique for a timer_t in OsTimer, if this handle is recycled, the descriptor needs
         /// to be incremented first
         std::atomic<uint32_t> m_descriptor{0u};
-        std::atomic_flag m_callbackIsAboutToBeExecuted;
+        // must be operator= otherwise it is undefined, see https://en.cppreference.com/w/cpp/atomic/ATOMIC_FLAG_INIT
+        std::atomic_flag m_callbackIsAboutToBeExecuted = ATOMIC_FLAG_INIT;
 
         std::atomic<bool> m_inUse{false};
         std::atomic<bool> m_isTimerActive{false};

--- a/iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/timer.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/timer.hpp
@@ -124,7 +124,7 @@ class Timer
         /// @brief the descriptor is unique for a timer_t in OsTimer, if this handle is recycled, the descriptor needs
         /// to be incremented first
         std::atomic<uint32_t> m_descriptor{0u};
-        std::atomic_flag m_callbackIsAboutToBeExecuted{false};
+        std::atomic_flag m_callbackIsAboutToBeExecuted;
 
         std::atomic<bool> m_inUse{false};
         std::atomic<bool> m_isTimerActive{false};

--- a/iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/grp.hpp
+++ b/iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/grp.hpp
@@ -21,10 +21,10 @@
 
 struct group
 {
-    char* gr_name;
-    char* gr_passwd;
+    const char* gr_name;
+    const char* gr_passwd;
     gid_t gr_gid;
-    char** gr_mem;
+    const char** gr_mem;
 };
 
 inline gid_t getegid(void)
@@ -34,8 +34,8 @@ inline gid_t getegid(void)
 
 inline struct group* getgrnam(const char* name)
 {
-    static char* groupName = "iceoryx_windows_group";
-    static char* groupPasswd = "iceoryx_windows_passwd";
+    static const char* groupName = "iceoryx_windows_group";
+    static const char* groupPasswd = "iceoryx_windows_passwd";
     static struct group dummy;
     dummy.gr_name = groupName;
     dummy.gr_passwd = groupPasswd;
@@ -46,8 +46,8 @@ inline struct group* getgrnam(const char* name)
 
 inline struct group* getgrgid(gid_t gid)
 {
-    static char* groupName = "iceoryx_windows_group";
-    static char* groupPasswd = "iceoryx_windows_passwd";
+    static const char* groupName = "iceoryx_windows_group";
+    static const char* groupPasswd = "iceoryx_windows_passwd";
     static struct group dummy;
     dummy.gr_name = groupName;
     dummy.gr_passwd = groupPasswd;

--- a/iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/pwd.hpp
+++ b/iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/pwd.hpp
@@ -21,19 +21,19 @@
 
 struct passwd
 {
-    char* pw_name;
-    char* pw_passwd;
+    const char* pw_name;
+    const char* pw_passwd;
     uid_t pw_uid;
     gid_t pw_gid;
-    char* pw_gecos;
-    char* pw_dir;
-    char* pw_shell;
+    const char* pw_gecos;
+    const char* pw_dir;
+    const char* pw_shell;
 };
 
 
 inline struct passwd* getpwnam(const char* name)
 {
-    static char* value = "iceoryx_windows_dummy";
+    static const char* value = "iceoryx_windows_dummy";
     static struct passwd dummy;
     dummy.pw_name = value;
     dummy.pw_passwd = value;
@@ -47,7 +47,7 @@ inline struct passwd* getpwnam(const char* name)
 
 inline struct passwd* getpwuid(uid_t uid)
 {
-    static char* value = "iceoryx_windows_dummy";
+    static const char* value = "iceoryx_windows_dummy";
     static struct passwd dummy;
     dummy.pw_name = value;
     dummy.pw_passwd = value;

--- a/iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/win32_errorHandling.hpp
+++ b/iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/win32_errorHandling.hpp
@@ -48,9 +48,9 @@ struct Win32CallReturn<void>
 
 template <typename Function, typename... Targs>
 auto __Win32Call(const char* functionName, const char* file, const int line, const Function& f, Targs&&... args)
-    -> Win32CallReturn<std::result_of_t<decltype(f)&(Targs...)>>
+    -> Win32CallReturn<std::invoke_result_t<decltype(f)&, Targs...>>
 {
-    Win32CallReturn<std::result_of_t<decltype(f)&(Targs...)>> retVal;
+    Win32CallReturn<std::invoke_result_t<decltype(f)&, Targs...>> retVal;
     SetLastError(0);
     retVal.assignValue(f, std::forward<Targs>(args)...);
     retVal.error = __PrintLastErrorToConsole(functionName, file, line);

--- a/iceoryx_hoofs/platform/win/source/semaphore.cpp
+++ b/iceoryx_hoofs/platform/win/source/semaphore.cpp
@@ -160,9 +160,9 @@ static HANDLE sem_create_win32_semaphore(LONG value, LPCSTR name)
     Win32Call(InitializeSecurityDescriptor, &securityDescriptor, SECURITY_DESCRIPTOR_REVISION).value;
 
     auto permissions = TEXT("D:") TEXT("(A;OICI;GA;;;BG)") // access to built-in guests
-        TEXT("(A;OICI;GA;;;AN)")                             // access to anonymous logon
-        TEXT("(A;OICI;GRGWGX;;;AU)")                         // access to authenticated users
-        TEXT("(A;OICI;GA;;;BA)");                            // access to administrators
+        TEXT("(A;OICI;GA;;;AN)")                           // access to anonymous logon
+        TEXT("(A;OICI;GRGWGX;;;AU)")                       // access to authenticated users
+        TEXT("(A;OICI;GA;;;BA)");                          // access to administrators
 
     Win32Call(ConvertStringSecurityDescriptorToSecurityDescriptor,
               reinterpret_cast<LPCSTR>(permissions),

--- a/iceoryx_hoofs/platform/win/source/semaphore.cpp
+++ b/iceoryx_hoofs/platform/win/source/semaphore.cpp
@@ -159,7 +159,7 @@ static HANDLE sem_create_win32_semaphore(LONG value, LPCSTR name)
     SECURITY_DESCRIPTOR securityDescriptor;
     Win32Call(InitializeSecurityDescriptor, &securityDescriptor, SECURITY_DESCRIPTOR_REVISION).value;
 
-    TCHAR* permissions = TEXT("D:") TEXT("(A;OICI;GA;;;BG)") // access to built-in guests
+    auto permissions = TEXT("D:") TEXT("(A;OICI;GA;;;BG)") // access to built-in guests
         TEXT("(A;OICI;GA;;;AN)")                             // access to anonymous logon
         TEXT("(A;OICI;GRGWGX;;;AU)")                         // access to authenticated users
         TEXT("(A;OICI;GA;;;BA)");                            // access to administrators

--- a/iceoryx_hoofs/platform/win/source/win32_errorHandling.cpp
+++ b/iceoryx_hoofs/platform/win/source/win32_errorHandling.cpp
@@ -1,5 +1,5 @@
 // Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
-// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/iceoryx_hoofs/source/posix_wrapper/timer.cpp
+++ b/iceoryx_hoofs/source/posix_wrapper/timer.cpp
@@ -363,7 +363,7 @@ cxx::expected<TimerError> Timer::OsTimer::restart(const units::Duration timeToWa
 
     if (gettimeResult.has_error())
     {
-        return gettimeResult;
+        return cxx::error<TimerError>(gettimeResult.get_error());
     }
 
     // Set new timeToWait value

--- a/iceoryx_hoofs/test/moduletests/test_ipc_channel.cpp
+++ b/iceoryx_hoofs/test/moduletests/test_ipc_channel.cpp
@@ -28,6 +28,7 @@ namespace
 using namespace ::testing;
 using namespace iox;
 using namespace iox::posix;
+using namespace iox::units::duration_literals;
 
 #if defined(__APPLE__)
 using IpcChannelTypes = Types<UnixDomainSocket>;

--- a/iceoryx_hoofs/test/moduletests/test_unit_duration.cpp
+++ b/iceoryx_hoofs/test/moduletests/test_unit_duration.cpp
@@ -22,6 +22,7 @@ namespace
 {
 using namespace ::testing;
 using namespace iox::units;
+using namespace iox::units::duration_literals;
 
 constexpr uint64_t SECONDS_PER_MINUTE = Duration::SECS_PER_MINUTE;
 constexpr uint64_t SECONDS_PER_HOUR = Duration::SECS_PER_HOUR;

--- a/iceoryx_posh/source/roudi/memory/memory_provider.cpp
+++ b/iceoryx_posh/source/roudi/memory/memory_provider.cpp
@@ -81,7 +81,7 @@ cxx::expected<MemoryProviderError> MemoryProvider::create() noexcept
 
     if (memoryResult.has_error())
     {
-        return memoryResult;
+        return cxx::error<MemoryProviderError>(memoryResult.get_error());
     }
 
     m_memory = memoryResult.value();

--- a/iceoryx_posh/test/integrationtests/test_mq_interface_startup_race.cpp
+++ b/iceoryx_posh/test/integrationtests/test_mq_interface_startup_race.cpp
@@ -35,6 +35,7 @@ using namespace ::testing;
 using namespace iox;
 using namespace iox::units;
 using namespace iox::posix;
+using namespace iox::units::duration_literals;
 
 using iox::runtime::IpcInterfaceBase;
 using iox::runtime::IpcMessage;

--- a/iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp
+++ b/iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp
@@ -16,6 +16,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "iceoryx_hoofs/cxx/convert.hpp"
+#include "iceoryx_hoofs/cxx/type_traits.hpp"
 #include "iceoryx_hoofs/cxx/unique_ptr.hpp"
 #include "iceoryx_posh/internal/mepoo/mem_pool.hpp"
 #include "iceoryx_posh/internal/mepoo/memory_manager.hpp"
@@ -141,7 +142,7 @@ TEST(ChunkHeader_test, ChunkHeaderBinaryCompatibilityCheck)
 TEST(ChunkHeader_test, ChunkHeaderUserPayloadSizeTypeIsLargeEnoughForMempoolChunk)
 {
     ::testing::Test::RecordProperty("TEST_ID", "540e2e95-0890-4522-ae7f-c6d867679e0b");
-    using ChunkSize_t = std::result_of<decltype (&MemPool::getChunkSize)(MemPool)>::type;
+    using ChunkSize_t = iox::cxx::invoke_result<decltype(&MemPool::getChunkSize), MemPool>::type;
 
     auto maxOfChunkSizeType = std::numeric_limits<ChunkSize_t>::max();
     auto maxOfUserPayloadSizeType = std::numeric_limits<decltype(std::declval<ChunkHeader>().userPayloadSize())>::max();

